src/benchq/resource_estimation/graph_compilation.py

```diff
@@ -189,29 +239,151 @@ def get_resource_estimations_for_program(
-        n_nodes_list,
+        graphs_list,
+        data_qubits_map_list,
         quantum_program,
         architecture_model,
         ec_error_budget,
+        use_full_program_graph=use_full_program_graph,
+        verbose=verbose,
     )
 
 
 def resource_estimations_for_subcomponents(
-    n_nodes_list, quantum_program, architecture_model, tolerable_circuit_error_rate
+    graphs_list,
+    data_qubits_map_list,
+    quantum_program,
+    architecture_model,
+    tolerable_circuit_error_rate,
+    use_full_program_graph: bool = False,
+    verbose: bool = False,
 ):
-    total_n_nodes = sum(
-        n_nodes * mult
-        for n_nodes, mult in zip(n_nodes_list, quantum_program.multiplicities)
+    n_nodes = sum(
+        len(graph) * multiplicity
+        for graph, multiplicity in zip(graphs_list, quantum_program.multiplicities)
     )
-
-    resource_estimates = get_resource_estimations_for_graph(
-        total_n_nodes, architecture_model, tolerable_circuit_error_rate
+    # account for double counting of data qubits coming from each graph
+    n_nodes -= quantum_program.num_data_qubits * (
+        len(quantum_program.subroutine_sequence) - 1
     )
-    resource_estimates["n_nodes"] = n_nodes_list
+
+    if use_full_program_graph:
+        program_graph = combine_subcomponent_graphs(
+            graphs_list, data_qubits_map_list, quantum_program
+        )
+        resource_estimates = get_resource_estimations_for_graph(
+            program_graph, architecture_model, tolerable_circuit_error_rate, verbose
+        )
+    else:
+        # use dummy graph
+        resource_estimates = get_resource_estimations_for_graph(
+            nx.path_graph(n_nodes),
+            architecture_model,
+            tolerable_circuit_error_rate,
+            verbose=False,
+        )
+        resource_estimates = get_substrate_scheduler_estimates_for_subcomponents(
+            graphs_list,
+            quantum_program,
+            data_qubits_map_list,
+            resource_estimates,
+            verbose,
+        )
+
+    resource_estimates["n_nodes"] = n_nodes
     resource_estimates["steps"] = quantum_program.steps
 
     return resource_estimates
+
+
+def combine_subcomponent_graphs(
+    graphs_list: List[nx.Graph],
+    data_qubits_map_list: List[List[int]],
+    quantum_program: QuantumProgram,
+):
+    """Given a list of graphs for each of the subcomponents, combine those graphs
+    into single graph representing the entire quantum program.
+
+    Args:
+        graphs_list (List[nx.Graph]): _description_
+        data_qubits_map_list (List[List[int]]): _description_
+        quantum_program (QuantumProgram): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    program_graph = graphs_list[quantum_program.subroutine_sequence[0]]
+    node_relabeling = {}
+    prev_graph_size = 0
+    for prev, curr in more_itertools.windowed(quantum_program.subroutine_sequence, 2):
```

Comment:
[nitpick] Would be easier to understand if we had sth like `prev_sequence`, `prev_graph` or sth like that,

Reply:
We talked about this.